### PR TITLE
Remove Policies interface

### PR DESCRIPTION
**Pull request by c633**
Remove `Policies` interface and use `ConiksPolicies` struct instead.

### CODE REVIEW

**Comment by arlolra** [comment]
Looks good, but are there any other policies besides CONIKS policies?  If not, maybe just call this `Policies`?

**Comment by c633** [comment]
Yes, it makes more sense.

File: merkletree/str.go [hunk 1]
```diff
@@ -21,10 +21,10 @@ type SignedTreeRoot struct {
 	PreviousEpoch   uint64
 	PreviousSTRHash []byte
 	Signature       []byte
-	Policies        Policies
+	Policies        *Policies
```
Comment (arlolra) [inline merkletree/str.go]: This is a late night comment, so take it for what it's worth.  Is there any chance that changing this to a pointer is going to lead to the Policies in the STR being mutated?

Comment (c633) [inline merkletree/str.go]: I think it's still safe by far. Changing this to pointer absolutely makes the Policies being mutated but it's the same for other fields of slice type.

Comment (arlolra) [inline merkletree/str.go]: Sort of ... there's an expectation that policies are to be able to be altered between epochs.  With this change, we're asking developers to be aware that they need to allocate a new Policies structure each time, rather than manipulate the one they may be retaining since startup. It just feels like a copy is less error prone.

After writing that sentence, I looked at directory.go and saw that we're doing this right,
https://github.com/coniks-sys/coniks-go/blob/master/protocol/directory.go#L43
but still feel uncomfortable with the pointer.  Maybe others have an opinion?

Also, I think we have a bug #80

Comment (c633) [inline merkletree/str.go]: > With this change, we're asking developers to be aware that they need to allocate a new Policies structure each time, rather than manipulate the one they may be retaining since startup

I don't think that we're asking developers to allocate a new Policies struct each time. They just need to allocate a new struct when the policies change, much like when we using the interface. 
@masomel @Liamsi : What do you think?

Comment (Liamsi) [inline merkletree/str.go]: Currently, the policies are accessed in the `merkletree` package only. Couldn't we even make the field unexported? So we hide the inner workings and still use a pointer here (which was introduced for performance reasons, I guess?)

Comment (c633) [inline merkletree/str.go]: Unfortunately, we need to export Policies's fields for json marshalling.

Comment (Liamsi) [inline merkletree/str.go]: Thank for the clarification. Then I'm also slightly in favour of making this a copy instead of a pointer. Simply because it is slightly more explicit.

**Comment by Liamsi** [comment]
LGTM (no strong opinion on this discussion: https://github.com/coniks-sys/coniks-go/pull/78#discussion_r77117504)

File: merkletree/str.go (hunk 1 above)
Comment (masomel) [inline merkletree/str.go]: I think I agree with @arlolra and @Liamsi, a copy seems safer than a pointer.

Comment (c633) [inline merkletree/str.go]: I fixed it in ceb0a58cdd139cb212c07dccabfb2f07be51520a. But a part of me still feels uncomfortable when considering the memory use between a copy and a pointer, in this case.

Maybe another solution is to ask developers to be aware that they need to allocate a new Policies structure each time, in the documentations.

Comment (arlolra) [inline merkletree/str.go]: Isn't the overhead massively dominated by the cloned trees in memory?  The tradeoff is the potential for a corrupted state, which seems worth it.

Unfortunately, documenting that sort of expectation is unreliable and I would caution to err on the side of programming defensively.

Comment (c633) [inline merkletree/str.go]: > Unfortunately, documenting that sort of expectation is unreliable and I would caution to err on the side of programming defensively.

Yup, I will take it as a lesson. Thanks!

**Comment by masomel** [comment]
> Btw, what do you think about ceb0a58? 

LGTM.

File: merkletree/pad.go [hunk 2]
```diff
@@ -21,12 +21,12 @@ type PAD struct {
 	snapshots    map[uint64]*SignedTreeRoot
 	loadedEpochs []uint64 // slice of epochs in snapshots
 	latestSTR    *SignedTreeRoot
-	policies     Policies // the current policies in place
+	policies     *Policies // the current policies in place
```
Comment (arlolra) [inline merkletree/pad.go]: `pad.policies` is in use, so the same mutation comment applies here.

I think you should revert everything in this patch that isn't strictly about removing the interface, sorry.

Comment (c633) [inline merkletree/pad.go]: > pad.policies is in use, so the same mutation comment applies here.

Yup, that's why I have this bug bc36c45e36c9eb51bc7510034ac43f80c44f3b30 (fixed in this commit). Thanks!

> I think you should revert everything in this patch that isn't strictly about removing the interface, sorry.

Oops, my bad. I split it into a separate commit (af994216261da5c00278929e1638e3ca682d4247)

Btw, I'm so sorry because of these late commits and response.

File: protocol/directory.go [hunk 3]
```diff
@@ -39,12 +39,15 @@ func (d *ConiksDirectory) Update() {
...
 
-func (d *ConiksDirectory) SetPolicies(epDeadline merkletree.TimeStamp, vrfKey vrf.PrivateKey) {
-	d.policies = merkletree.NewPolicies(epDeadline, vrfKey)
+func (d *ConiksDirectory) UpdatePolicies(epDeadline merkletree.TimeStamp, vrfKey vrf.PrivateKey) {
+	d.policies.Update(epDeadline, vrfKey)
 }
 
+// EpochDeadline returns the EpochDeadline of the current policies
+// If you want to query the EpochDeadline before the new policies
+// is updated, use LatestSTR().Policies.EpochDeadline
```
Comment (arlolra) [inline protocol/directory.go]: But that's exactly what you want to use in `EpochUpdate`.

https://github.com/coniks-sys/coniks-go/blob/master/keyserver/server.go#L192

The timer needs to fire based on `LatestSTR`'s deadline.

Again, I really think you should revert everything here that isn't about removing the interface.  The new test is good to add, but orthogonal.

Comment (c633) [inline protocol/directory.go]: > Again, I really think you should revert everything here that isn't about removing the interface. The new test is good to add, but orthogonal.

I didn't get your idea yet :( What should be removed?

Comment (arlolra) [inline protocol/directory.go]: Sorry for not being clear. Just the `UpdatePolicies` stuff.  The shallow copy in `pad.policies = *policies` is all we were after.

Let me update this, one sec.

**Comment by arlolra** [comment]
@c633 See d0aff9a.  If you're happy with that, please merge.  Thanks!

**Comment by c633** [comment]
Thanks!
